samples/custom_dispatch/cpu/embedded/udon_spec.mlir

```diff
@@ -0,0 +1,173 @@
+
+    // Dispatch a basic `ret = |lhs * rhs|` using an external function.
+    %0 = flow.dispatch @executable::@x86_64::@simple_mul_abs_negate[%dim](%dim_i32, %arg0, %arg1) {
+      hal.interface.bindings = [
+        #hal.interface.binding<0, 0>,
+        #hal.interface.binding<0, 1>,
+        #hal.interface.binding<0, 2>
+      ],
+      // HACK: keep the executable live through DCE. Only required when
+      // using the automatic variant selection.
+      hal.executable.ref = [@executable]
+    } : (i32, tensor<?xf32>{%dim}, tensor<?xf32>{%dim}) -> tensor<?xf32>{%dim}
+    return %0 : tensor<?xf32>
+  }
+
+  transform.named_sequence @match_mul_abs_negate(%root: !transform.any_op {transform.readonly}) -> (!transform.any_value, !transform.any_value) {
+    %ins, %outs = transform.iree.match.cast_compatible_dag_from_root %root {
+      ^bb0(%lhs: tensor<?xf32>, %rhs: tensor<?xf32>):
+        // The matcher does not recurse to the constant index + dim because
+        // their only consumer matches only the operation name.
+        %c0 = arith.constant 0 : index
+        %dim = tensor.dim %lhs, %c0 : tensor<?xf32>
+        // --------------------------------------------------------------------
+        %empty = tensor.empty(%dim) {"match.operation_name_only"} : tensor<?xf32>
+        %mul = linalg.generic {indexing_maps = [affine_map<(d0) -> (d0)>,
+                                                affine_map<(d0) -> (d0)>,
+                                                affine_map<(d0) -> (d0)>],
+                               iterator_types = ["parallel"]}
+                               ins(%lhs, %rhs : tensor<?xf32>, tensor<?xf32>)
+                               outs(%empty : tensor<?xf32>) {
+        ^bb0(%in: f32, %in0: f32, %out: f32):
+          %m = arith.mulf %in, %in0 : f32
+          linalg.yield %m : f32
+        } -> tensor<?xf32>
+        %abs = linalg.generic {indexing_maps = [affine_map<(d0) -> (d0)>,
+                                                affine_map<(d0) -> (d0)>],
+                               iterator_types = ["parallel"]}
+                               ins(%mul : tensor<?xf32>)
+                               outs(%empty : tensor<?xf32>) {
+        ^bb0(%in: f32, %out: f32):
+          %a = math.absf %in : f32
+          linalg.yield %a : f32
+        } -> tensor<?xf32>
+        %neg = linalg.generic {indexing_maps = [affine_map<(d0) -> (d0)>,
+                                                affine_map<(d0) -> (d0)>],
+                               iterator_types = ["parallel"]}
+                               ins(%abs : tensor<?xf32>)
+                               outs(%empty : tensor<?xf32>) {
+        ^bb0(%in: f32, %out: f32):
+          %n = arith.negf %in : f32
+          linalg.yield %n : f32
+        } -> tensor<?xf32>
+    } : (!transform.any_op) -> (!transform.any_value, !transform.any_value)
+    transform.yield %ins, %outs : !transform.any_value, !transform.any_value
+  }
```

Comment:
@MaheshRavishankar this is almost ready for review, just need to wrap up some tests and finish some documentation, but I added this example with a more complex matcher (and hopefully a little less verbose, it just does a structural comparison). We will still need some work on the Torch side if we want to do matching there because of the different tensor type.

Reply:
Awesome! This has grown a lot since I last saw it... Ill start reviewing.

Reply:
ok this is ready for review now. The major additions are
 - The samples in `samples/custom_dispatch`
 - The transform extensions in Preprocessing for matching (+ the interpreter pass)
 
 The matching extensions are a bit experimental and we can change them as we go, but hopefully they provide a good enough starting point.